File: src/LeetCode/Q209.cpp
// 209. Minimum Size Subarray Sum
// Given an array of positive integers nums and a positive integer target, return the minimal length of a contiguous subarray [numsl, numsl+1, ..., numsr-1, numsr] of which the sum is greater than or equal to target. If there is no such subarray, return 0 instead.

#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:

    // original
    // sliding window
    // not very efficient
    int minSubArrayLen(int target, vector<int>& nums) {
        int n = nums.size();
        int result = 0;
        int b = 0, e = 0;
        // sum of nums[b:e]
        int sum = 0;
        // keep nums[b:e] a window
        do {
            if (sum >= target) {
                result = result == 0 ? e - b : min(result, e - b);
                sum -= nums[b++];
            } else {
                ++e;
                if (e <= n) {
                    sum += nums[e - 1];
                }
            }
        } while (b < e && e <= n);
        return result;
    }

    // original
    // sliding window
    // not much improvement
    int minSubArrayLen2(int target, vector<int>& nums) {
        int n = nums.size();
        int result = n + 1;
        int sum = 0;
        int b = 0, e = -1;
        while (e < n) {
            if (sum >= target) {
                result = e - b + 1 < result ? e - b + 1 : result;
                sum -= nums[b++];
            } else {
                ++e;
                if (e < n) {
                    sum += nums[e];
                }
            }
        }
        return result == n + 1 ? 0 : result;
    }
};

int main() {

    Solution s;
    vector<int> v1{2,3,1,2,4,3};
    cout << s.minSubArrayLen2(7, v1) << endl; // 2
    vector<int> v2{1, 4, 4};
    cout << s.minSubArrayLen2(4, v2) << endl; // 1
    vector<int> v3{1,1,1,1,1,1,1,1};
    cout << s.minSubArrayLen2(11, v3) << endl; // 0
    vector<int> v4{1, 2, 3, 4, 5};
    cout << s.minSubArrayLen2(11, v4) << endl; // 3
}

File: LeetCode/python/Q209.py
# 209. Minimum Size Subarray Sum
# Medium
# Given an array of positive integers nums and a positive integer target, return the minimal length
# of a subarray whose sum is greater than or equal to target. If there is no such subarray,
# return 0 instead.
from typing import List


# 2023-07-06 00:00:51
# original
# sliding window
class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        n = len(nums)
        start = 0
        ans = n + 1
        current_sum = 0
        for i in range(n):
            current_sum += nums[i]
            while current_sum >= target:
                ans = min(ans, i - start + 1)
                current_sum -= nums[start]
                start += 1
        return min(ans, n - start + 1) if ans != n + 1 else 0


s = Solution()
print(s.minSubArrayLen(7, [2, 3, 1, 2, 4, 3]))  # 2
print(s.minSubArrayLen(4, [1, 4, 4]))  # 1
print(s.minSubArrayLen(11, [1, 1, 1, 1, 1, 1, 1, 1]))  # 0
print(s.minSubArrayLen(80, [10, 5, 13, 4, 8, 4, 5, 11, 14, 9, 16, 10, 20, 8]))  # 6
